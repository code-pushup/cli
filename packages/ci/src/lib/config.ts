import type { MonorepoTool } from './monorepo';

export type CIConfigInput = {
  monorepo?: boolean | MonorepoTool;
  projects?: string[] | null;
  task?: string;
  bin?: string;
  config?: string | null;
  directory?: string;
  silent?: boolean;
  annotations?: boolean;
  logger: CILogger;
};

export type CIConfig = Required<CIConfigInput>;

export const CI_CONFIG_DEFAULTS = {
  monorepo: false,
  projects: null,
  task: 'code-pushup',
  bin: 'npx --no-install code-pushup',
  config: null,
  directory: process.cwd(),
  silent: false,
  annotations: true,
  logger: console,
} satisfies Partial<CIConfig>;

export type CILogger = {
  error: (message: string) => void;
  warn: (message: string) => void;
  info: (message: string) => void;
  debug: (message: string) => void;
};
