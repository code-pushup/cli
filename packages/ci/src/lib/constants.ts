import type { Settings } from './models.js';

export const DEFAULT_SETTINGS: Settings = {
  monorepo: false,
  parallel: false, // TODO: default to true once battle-tested?
  projects: null,
  task: 'code-pushup',
  bin: 'npx --no-install code-pushup',
  config: null,
  directory: process.cwd(),
  silent: false,
  debug: false,
  detectNewIssues: true,
  logger: console,
  nxProjectsFilter: '--with-target={task}',
};
