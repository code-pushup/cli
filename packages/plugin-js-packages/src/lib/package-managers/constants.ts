export const COMMON_AUDIT_ARGS = ['audit', '--json'];
export const COMMON_OUTDATED_ARGS = ['outdated', '--json'];
