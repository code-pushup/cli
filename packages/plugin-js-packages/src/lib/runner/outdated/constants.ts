import { IssueSeverity } from '@code-pushup/models';
import { VersionType } from './types';

export const outdatedSeverity: Record<VersionType, IssueSeverity> = {
  major: 'error',
  minor: 'warning',
  patch: 'info',
};
