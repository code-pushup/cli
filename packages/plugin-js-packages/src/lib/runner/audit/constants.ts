import { PackageAuditLevel } from '../../config';

/* eslint-disable no-magic-numbers */
export const auditScoreModifiers: Record<PackageAuditLevel, number> = {
  critical: 1,
  high: 0.1,
  moderate: 0.05,
  low: 0.02,
  info: 0.01,
};
/* eslint-enable no-magic-numbers */
