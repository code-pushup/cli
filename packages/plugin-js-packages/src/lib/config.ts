import { z } from 'zod';
import { IssueSeverity, issueSeveritySchema } from '@code-pushup/models';
import { defaultAuditLevelMapping } from './constants';

export const dependencyGroups = ['prod', 'dev', 'optional'] as const;
const dependencyGroupSchema = z.enum(dependencyGroups);
export type DependencyGroup = (typeof dependencyGroups)[number];

const packageCommandSchema = z.enum(['audit', 'outdated']);
export type PackageCommand = z.infer<typeof packageCommandSchema>;

const packageManagerIdSchema = z.enum([
  'npm',
  'yarn-classic',
  'yarn-modern',
  'pnpm',
]);
export type PackageManagerId = z.infer<typeof packageManagerIdSchema>;

export const packageAuditLevels = [
  'critical',
  'high',
  'moderate',
  'low',
  'info',
] as const;
const packageAuditLevelSchema = z.enum(packageAuditLevels);
export type PackageAuditLevel = z.infer<typeof packageAuditLevelSchema>;

export type AuditSeverity = Record<PackageAuditLevel, IssueSeverity>;

export function fillAuditLevelMapping(
  mapping: Partial<AuditSeverity>,
): AuditSeverity {
  return {
    critical: mapping.critical ?? defaultAuditLevelMapping.critical,
    high: mapping.high ?? defaultAuditLevelMapping.high,
    moderate: mapping.moderate ?? defaultAuditLevelMapping.moderate,
    low: mapping.low ?? defaultAuditLevelMapping.low,
    info: mapping.info ?? defaultAuditLevelMapping.info,
  };
}

export const jsPackagesPluginConfigSchema = z.object({
  checks: z
    .array(packageCommandSchema, {
      description:
        'Package manager commands to be run. Defaults to both audit and outdated.',
    })
    .min(1)
    .default(['audit', 'outdated']),
  packageManager: packageManagerIdSchema.describe(
    'Package manager to be used.',
  ),
  dependencyGroups: z
    .array(dependencyGroupSchema)
    .min(1)
    .default(['prod', 'dev']),
  auditLevelMapping: z
    .record(packageAuditLevelSchema, issueSeveritySchema, {
      description:
        'Mapping of audit levels to issue severity. Custom mapping or overrides may be entered manually, otherwise has a default preset.',
    })
    .default(defaultAuditLevelMapping)
    .transform(fillAuditLevelMapping),
  packageJsonPath: z
    .string()
    .describe('File path to package.json. Defaults to current folder.')
    .default('package.json'),
});

export type JSPackagesPluginConfig = z.input<
  typeof jsPackagesPluginConfigSchema
>;

export type FinalJSPackagesPluginConfig = z.infer<
  typeof jsPackagesPluginConfigSchema
>;
