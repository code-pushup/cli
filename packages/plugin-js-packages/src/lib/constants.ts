import { IssueSeverity } from '@code-pushup/models';
import type { DependencyGroup, PackageAuditLevel } from './config';
import { DependencyGroupLong } from './runner/outdated/types';

export const defaultAuditLevelMapping: Record<
  PackageAuditLevel,
  IssueSeverity
> = {
  critical: 'error',
  high: 'error',
  moderate: 'warning',
  low: 'warning',
  info: 'info',
};

export const dependencyGroupToLong: Record<
  DependencyGroup,
  DependencyGroupLong
> = {
  prod: 'dependencies',
  dev: 'devDependencies',
  optional: 'optionalDependencies',
};

/* eslint-disable no-magic-numbers */
export const dependencyGroupWeights: Record<DependencyGroup, number> = {
  prod: 80,
  dev: 15,
  optional: 5,
};
/* eslint-enable no-magic-numbers */

export const dependencyDocs: Record<DependencyGroup, string> = {
  prod: 'https://classic.yarnpkg.com/docs/dependency-types#toc-dependencies',
  dev: 'https://classic.yarnpkg.com/docs/dependency-types#toc-devdependencies',
  optional:
    'https://classic.yarnpkg.com/docs/dependency-types#toc-optionaldependencies',
};
