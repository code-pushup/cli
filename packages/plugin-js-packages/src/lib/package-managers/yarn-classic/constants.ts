import { OutdatedDependency } from '../../runner/outdated/types';
import { Yarnv1FieldName } from './types';

export const outdatedtoFieldMapper: Record<
  keyof OutdatedDependency,
  Yarnv1FieldName
> = {
  name: 'Package',
  current: 'Current',
  latest: 'Latest',
  type: 'Package Type',
  url: 'URL',
};

export const REQUIRED_OUTDATED_FIELDS: Yarnv1FieldName[] = [
  'Package',
  'Current',
  'Latest',
  'Package Type',
];
