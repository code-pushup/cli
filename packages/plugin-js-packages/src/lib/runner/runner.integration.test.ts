import { rm, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { MockInstance, describe, expect, it } from 'vitest';
import { AuditOutputs, RunnerConfig } from '@code-pushup/models';
import {
  ensureDirectoryExists,
  readJsonFile,
  removeDirectoryIfExists,
} from '@code-pushup/utils';
import { createRunnerConfig, executeRunner } from '.';
import { FinalJSPackagesPluginConfig } from '../config';
import { defaultAuditLevelMapping } from '../constants';
import { PLUGIN_CONFIG_PATH, RUNNER_OUTPUT_PATH, WORKDIR } from './constants';

describe('executeRunner', () => {
  let cwdSpy: MockInstance<[], string>;
  const appDir = join(
    fileURLToPath(dirname(import.meta.url)),
    '..',
    '..',
    '..',
    '..',
    '..',
    'examples',
    'react-todos-app',
  );

  beforeAll(async () => {
    await ensureDirectoryExists(dirname(PLUGIN_CONFIG_PATH));
    cwdSpy = vi.spyOn(process, 'cwd').mockReturnValue(appDir);
  });

  afterAll(async () => {
    cwdSpy.mockRestore();
    await rm(PLUGIN_CONFIG_PATH, { force: true });
  });

  it('should successfully execute runner for npm audit', async () => {
    const config: FinalJSPackagesPluginConfig = {
      packageManager: 'npm',
      checks: ['audit'],
      auditLevelMapping: defaultAuditLevelMapping,
    };
    await writeFile(PLUGIN_CONFIG_PATH, JSON.stringify(config));
    await executeRunner();
    const results = await readJsonFile<AuditOutputs>(RUNNER_OUTPUT_PATH);
    expect(results).toMatchSnapshot();
  });

  it('should successfully execute runnerfor npm outdated', async () => {
    const config: FinalJSPackagesPluginConfig = {
      packageManager: 'npm',
      checks: ['outdated'],
      auditLevelMapping: defaultAuditLevelMapping,
    };
    await writeFile(PLUGIN_CONFIG_PATH, JSON.stringify(config));
    await executeRunner();
    const results = await readJsonFile<AuditOutputs>(RUNNER_OUTPUT_PATH);
    expect(results).toMatchSnapshot();
  });
}, 10_000);

describe('createRunnerConfig', () => {
  it('should create a valid runner config', async () => {
    const runnerConfig = await createRunnerConfig('executeRunner.ts', {
      packageManager: 'npm',
      checks: ['audit'],
      auditLevelMapping: defaultAuditLevelMapping,
    });
    expect(runnerConfig).toStrictEqual<RunnerConfig>({
      command: 'node',
      args: ['executeRunner.ts'],
      outputFile: expect.stringContaining('runner-output.json'),
    });
  });

  it('should provide plugin config to runner in JSON file', async () => {
    await removeDirectoryIfExists(WORKDIR);
    const pluginConfig: FinalJSPackagesPluginConfig = {
      packageManager: 'yarn-classic',
      checks: ['outdated'],
      auditLevelMapping: { ...defaultAuditLevelMapping, moderate: 'error' },
    };
    await createRunnerConfig('executeRunner.ts', pluginConfig);
    const config = await readJsonFile<FinalJSPackagesPluginConfig>(
      PLUGIN_CONFIG_PATH,
    );
    expect(config).toStrictEqual(pluginConfig);
  });
});
