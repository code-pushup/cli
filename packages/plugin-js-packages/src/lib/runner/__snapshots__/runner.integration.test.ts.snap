// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`executeRunner > should successfully execute runner for npm audit 1`] = `
[
  {
    "displayValue": "0 vulnerabilities",
    "score": 1,
    "slug": "npm-audit-prod",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "\`verdaccio\`'s dependency \`jsonwebtoken\` has a **moderate** vulnerability in versions **<=8.5.1**. Fix available: Update \`verdaccio\` to version **5.29.2**. More information: [jsonwebtoken unrestricted key type could lead to legacy keys usage ](https://github.com/advisories/GHSA-8cf7-32gw-wr33)",
          "severity": "warning",
        },
        {
          "message": "\`verdaccio\`'s dependency \`request\` has a **moderate** vulnerability in **all** versions. Fix available: Update \`verdaccio\` to version **5.29.2**. More information: [Server-Side Request Forgery in Request](https://github.com/advisories/GHSA-p8p7-x288-28g6)",
          "severity": "warning",
        },
        {
          "message": "\`verdaccio\`'s dependency \`semver\` has a **moderate** vulnerability in versions **7.0.0 - 7.5.1**. Fix available: Update \`verdaccio\` to version **5.29.2**. More information: [semver vulnerable to Regular Expression Denial of Service](https://github.com/advisories/GHSA-c2qf-rxjj-qqgw)",
          "severity": "warning",
        },
        {
          "message": "\`request\`'s dependency \`tough-cookie\` has a **moderate** vulnerability in versions **<4.1.3**. Fix available: Update \`verdaccio\` to version **5.29.2**. More information: [tough-cookie Prototype Pollution vulnerability](https://github.com/advisories/GHSA-72xf-g2v4-qvf3)",
          "severity": "warning",
        },
        {
          "message": "\`verdaccio\` dependency has a **moderate** vulnerability in versions **<=5.28.0**. Fix available: Update \`verdaccio\` to version **5.29.2**. More information: [semver vulnerable to Regular Expression Denial of Service](https://github.com/advisories/GHSA-c2qf-rxjj-qqgw)",
          "severity": "warning",
        },
      ],
    },
    "displayValue": "5 vulnerabilities (5 moderate)",
    "score": 0.75,
    "slug": "npm-audit-dev",
    "value": 5,
  },
  {
    "displayValue": "0 vulnerabilities",
    "score": 1,
    "slug": "npm-audit-optional",
    "value": 0,
  },
]
`;

exports[`executeRunner > should successfully execute runnerfor npm outdated 1`] = `
[
  {
    "displayValue": "all dependencies are up to date",
    "score": 1,
    "slug": "npm-outdated-prod",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Package [\`memfs\`](https://github.com/streamich/memfs) requires a **minor** update from **4.6.1** to **4.8.0**.",
          "severity": "warning",
        },
        {
          "message": "Package [\`prettier\`](https://prettier.io) requires a **major** update from **2.8.8** to **3.2.5**.",
          "severity": "error",
        },
        {
          "message": "Package [\`vite\`](https://vitejs.dev) requires a **major** update from **4.5.2** to **5.1.6**.",
          "severity": "error",
        },
        {
          "message": "Package [\`vitest\`](https://github.com/vitest-dev/vitest#readme) requires a **patch** update from **1.3.0** to **1.3.1**.",
          "severity": "info",
        },
      ],
    },
    "displayValue": "2 out of 4 outdated dependencies require major update",
    "score": 0.6666666666666666,
    "slug": "npm-outdated-dev",
    "value": 4,
  },
  {
    "displayValue": "all dependencies are up to date",
    "score": 1,
    "slug": "npm-outdated-optional",
    "value": 0,
  },
]
`;
